test_conformance/math_brute_force/unary_two_results_half.cpp

```diff
@@ -0,0 +1,459 @@
+//
+// Copyright (c) 2023 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "common.h"
+#include "function_list.h"
+#include "test_functions.h"
+#include "utility.h"
+
+#include <cinttypes>
+#include <cstring>
+
+namespace {
+
+cl_int BuildKernelFn_HalfFn(cl_uint job_id, cl_uint thread_id UNUSED, void *p)
+{
+    BuildKernelInfo &info = *(BuildKernelInfo *)p;
+    auto generator = [](const std::string &kernel_name, const char *builtin,
+                        cl_uint vector_size_index) {
+        return GetUnaryKernel(kernel_name, builtin, ParameterType::Half,
+                              ParameterType::Half, ParameterType::Half,
+                              vector_size_index);
+    };
+    return BuildKernels(info, job_id, generator);
+}
+
+} // anonymous namespace
+
+int TestFunc_Half2_Half(const Func *f, MTdata d, bool relaxedMode)
+{
+    int error;
+    Programs programs;
+    const unsigned thread_id = 0; // Test is currently not multithreaded.
+    KernelMatrix kernels;
+    float maxError0 = 0.0f;
+    float maxError1 = 0.0f;
+    int ftz = f->ftz || gForceFTZ || 0 == (CL_FP_DENORM & gHalfCapabilities);
+    float maxErrorVal0 = 0.0f;
+    float maxErrorVal1 = 0.0f;
+    uint64_t step = getTestStep(sizeof(cl_half), BUFFER_SIZE);
+
+    constexpr size_t half_buffer_size = BUFFER_SIZE / sizeof(cl_half);
+
+    cl_uchar overflow[half_buffer_size];
+    int isFract = 0 == strcmp("fract", f->nameInCode);
```

Comment:
I'm seeing a verification failure for fract with a very small input value:
```
ERROR: fract2: {-0.874878, 0.000000} ulp error at -0x1.004p-14: *{0x1p+0, -0x1p+0} vs. {0x1.ffcp-1, -0x1p+0}
```

I believe the reference calculation of the return value (`0x1p+0`) is wrong, as the OpenCL C spec s7.5.1 says:
> fract(x, iptr) shall not return a value greater than or equal to 1.0,

Reply:
Indeed, input `-0x1.004p-14` (`-6.109476e-05`, hex `0x8401`) returns `0.999938905` reference float which is converted into `1.0` half float due to default `RTE` rounding method. 

Still, according to `cl_fp16_khr` extension spec `fract` should return correctly rounded result which is `1.0`, right ?

If I will put additional condition to replace reference value `1.0` with largest number less than one (`0x3bff`) my intel implementation does not pass the test (kernel `fract` returns `1.0`).

Reply:
> Still, according to cl_fp16_khr extension spec fract should return correctly rounded result which is 1.0, right ?

It doesn't say explicitly when the rounding should happen, but doing the rounding after the `fmin` step violates s7.5.1.  Hence I would argue that the intermediate result (i.e. `x - floor(x)`) should be rounded, so that the `fmin` step only sees correctly rounded fp16 values.

Intuitively I would expect `fract` to never return `1.0`, as its name suggests it returns the fractional part of the input. The purpose of the `fmin` operator is to enforce that, by capping the return value to `0x1.ffcp-1` (the largest fp16 value less than `1.0`).

Reply:
>If I will put additional condition to replace reference value 1.0 with largest number less than one (0x3bff) my intel implementation does not pass the test (kernel `fract` returns 1.0).

@bashbaug could you be so kind to comment on that ? Thanks

Reply:
I think this is an issue with our `fract` implementation.  I just filed https://github.com/intel/intel-graphics-compiler/issues/307.  Thanks!

Reply:
Thanks, correction applied.

Reply:
Thanks, latest correction fixes the `fract` issue.